chia/server/upnp.py

```diff
@@ -7,16 +8,41 @@
 
 
 log = logging.getLogger(__name__)
+upnp: miniupnpc.UPnP = None
 
 
-def upnp_remap_port(port) -> None:
-    log.info(f"Attempting to enable UPnP (open up port {port})")
+def setup_upnp():
     try:
+        global upnp
         upnp = miniupnpc.UPnP()
         upnp.discoverdelay = 30
         upnp.discover()
         upnp.selectigd()
+    except Exception as e:
+        log.info("UPnP failed. This is not required to run chia, but it allows incoming connections from other peers.")
+        log.info(e)
+
+
+threading.Thread(target=setup_upnp)
```

Comment:
does this create a detached thread?
when the program shuts down, there's no way to wait on this thread completing.

I think you need to get a handle to this thread so we can shut it down gracefully and wait/join it on shutdown, but also so we can post more upnp jobs to it.